### PR TITLE
ROX-24840 Docker client API for integration tests

### PR DESCRIPTION
**Pull request by robbycochran**
## Description

Refactor integration tests from using docker process execution to using the docker engine API. 
- Decouple k8s integration tests from container test actions (this made it easier to refactor). `K8sNamespaceTestSuite` no longer inherits from the `BaseIntegrationTestSuite`.
- Remove all shell execution from test suite code -- all system interactions are through Executor interface. 
- Launch all containers using unified configuration (`ContainerStartConfig`) rather than command line args 
- Remove Docker process executor
- Add `-dockerized` test targets in the integration-test makefile, e.g. `make -C integration-tests TestProcessNetwork-dockerized`
- Adds [logging](https://github.com/stackrox/collector/actions/runs/10623486844?pr=1792) to provide visibility and help with debugging issues. Also print last few lines of collector logs when there is a non-zero exit.   

### Logging example
```
2024/08/29 23:20:50 INFO: read credentials for quay.io from /root/.docker/config.json
2024/08/29 23:20:51 INFO: registry login success: quay.io
2024/08/29 23:20:51 INFO: quay.io/rhacs-eng/collector-performance:stats-1.1.3 already exists
2024/08/29 23:20:51 INFO: start container-stats with quay.io/rhacs-eng/collector-performance:stats-1.1.3 (19a07c711590)
2024/08/29 23:20:51 INFO: pulling quay.io quay.io/rhacs-eng/collector:3.19.x-73-g3523d378a9
2024/08/29 23:21:10 INFO: pulled quay.io/rhacs-eng/collector:3.19.x-73-g3523d378a9
2024/08/29 23:21:13 INFO: start collector with quay.io/rhacs-eng/collector:3.19.x-73-g3523d378a9 (f3a2749e296c)
2024/08/29 23:21:13 INFO: collector has healthcheck: CMD-SHELL /usr/local/bin/status-check.sh
2024/08/29 23:22:13 ERROR: Waiting for container collector to become health=healthy, elapsed time: 1m0.002452957s
2024/08/29 23:22:13 ERROR: Timed out waiting for container collector to become health=healthy, elapsed Time: 

...

    	            	collector AbortHandler 0xa6d161 + 38
    	            	/lib64/libc.so.6 (null) 0x7dabbca2a6f0 + 0
    	            	/lib64/libc.so.6 (null) 0x7dabbca7794c + 0
    	            	/lib64/libc.so.6 raise 0x7dabbca2a646 + 22
    	            	/lib64/libc.so.6 abort 0x7dabbca147f3 + 211
    	            	/lib64/libstdc++.so.6 (null) 0x7dabbcd8cb21 + 0
    	            	/lib64/libstdc++.so.6 (null) 0x7dabbcd9852c + 0
    	            	/lib64/libstdc++.so.6 (null) 0x7dabbcd98597 + 0
    	            	/lib64/libstdc++.so.6 (null) 0x7dabbcd987f9 + 0
    	            	collector collector::KernelDriverCOREEBPF::Setup(collector::CollectorConfig const&amp;, sinsp&amp;) 0xb1b1c9 + 303
    	            	collector collector::system_inspector::Service::InitKernel(collector::CollectorConfig const&amp;) 0xb17d98 + 1530
    	            	collector collector::CollectorService::InitKernel() 0xa7dc62 + 38
    	            	collector collector::SetupKernelDriver(collector::CollectorService&amp;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;, collector::CollectorConfig const&amp;) 0xa7de08 + 159
    	            	collector main 0xa69e1d + 1269
    	            	/lib64/libc.so.6 (null) 0x7dabbca15590 + 0
    	            	/lib64/libc.so.6 __libc_start_main 0x7dabbca15640 + 128
    	            	collector _start 0xa68c65 + 37
    	            	Caught signal 6 (SIGABRT): Aborted
    	            	/bootstrap.sh: line 85:     9 Aborted                 eval exec &quot;$@&quot;
    	            	
    	Test:       	TestProcessNetwork</code></pre></td></tr>
```

## Checklist
- [x] Investigated and inspected CI test results
- [x] Updated documentation accordingly

**Automated testing**
  - [x] Added integration tests
~- [ ] Added unit tests~
~- [ ] Added regression tests~

If any of these don't apply, please comment below.

## Testing Performed
checked logs

### CODE REVIEW

File: integration-tests/suites/k8s_namespace.go [hunk 1]
```diff
@@ -104,18 +134,20 @@ func (k *K8sNamespaceTestSuite) TearDownSuite() {
 }
 
 func (k *K8sNamespaceTestSuite) TestK8sNamespace() {
+	// Sleep to ensure asynchronous container engine lookups have completed
+	time.Sleep(10 * time.Second)
```
Comment (Molter73) [inline integration-tests/suites/k8s_namespace.go:138]: Why is this sleep needed now? Were you having timing issues when running locally?

**Review by Molter73 — COMMENTED** [review]
Quite a bit of comments, but nothing major. The change is looking great!

File: integration-tests/suites/k8s_namespace.go (hunk 1 above)
Comment (robbycochran) [reply integration-tests/suites/k8s_namespace.go:138]: Removed this sleep. I was having issues getting the tests to pass locally, but eventually got it working.

File: integration-tests/suites/base.go [hunk 2]
```diff
@@ -309,13 +314,66 @@ func (s *IntegrationTestSuiteBase) GetLogLines(containerName string) []string {
...
-
-	outLines := strings.Split(output, "\n")
-	return outLines[len(outLines)-1], err
+	if len(args) == 0 {
+		return "", errors.New("no arguments")
+	}
+	c := config.ContainerStartConfig{}
+	c.Name = name
+	skip := false
+	for i, arg := range args {
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/suites/base.go]: Removed this func and switched callsites to `StartContainer`

File: integration-tests/suites/base.go [hunk 3]
```diff
@@ -309,13 +314,66 @@ func (s *IntegrationTestSuiteBase) GetLogLines(containerName string) []string {
 }
 
 func (s *IntegrationTestSuiteBase) launchContainer(name string, args ...string) (string, error) {
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/suites/base.go]: went ahead and did it

File: integration-tests/pkg/executor/executor_docker_api.go [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+	if len(containerJSON.NetworkSettings.Ports) > 0 {
+		for portStr := range containerJSON.NetworkSettings.Ports {
+			portSplit := strings.Split(string(portStr), "/")
+			if len(portSplit) > 0 {
+				containerPort = portSplit[0]
+			}
+		}
+	}
+	log.Info("port for %s is %s\n", containerID, containerPort)
+	return containerPort, nil
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/executor/executor_docker_api.go:287]: done

File: integration-tests/pkg/executor/executor_docker_api.go [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+	}
+	containerPort := "-1"
+	if len(containerJSON.NetworkSettings.Ports) > 0 {
+		for portStr := range containerJSON.NetworkSettings.Ports {
+			portSplit := strings.Split(string(portStr), "/")
+			if len(portSplit) > 0 {
+				containerPort = portSplit[0]
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/executor/executor_docker_api.go]: yep, done.

File: integration-tests/pkg/executor/executor_docker_api.go [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+	log.Info("exec %s %v (exitCode=%d, outBytes=%d)\n",
+		containerName, command, execInspect.ExitCode, stdoutBuf.Len())
+	return stdoutBuf.String(), nil
+}
+
+func (d *dockerAPIExecutor) GetContainerLogs(containerID string) (string, error) {
+	ctx := context.Background()
+	defer d.client.Close()
+	timeoutContext, cancel := context.WithTimeout(ctx, 30*time.Second)
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/executor/executor_docker_api.go]: simplified the timeoutContext usage here and elsewhere.

File: integration-tests/pkg/executor/executor_docker_api.go [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+		return "", fmt.Errorf("container %s does not have a health check", containerID)
+	}
+
+	log.Info("%s has healthcheck: %s\n", containerID, strings.Join(inspectResp.Config.Healthcheck.Test, " "))
+	return strings.Join(inspectResp.Config.Healthcheck.Test, " "), nil
+}
+
+func (d *dockerAPIExecutor) StartContainer(startConfig config.ContainerStartConfig) (string, error) {
+	ctx := context.Background()
+	defer d.client.Close()
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/executor/executor_docker_api.go]: Yeah it isn't necessary. I don't think it closes the client, just closes the existing connections. I've removed calls to it.

File: integration-tests/pkg/executor/executor_docker_api.go [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+func newDockerAPIExecutor() (*dockerAPIExecutor, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, err
+	}
+
+	authConfigs := map[string]string{}
+	auths, err := readRegistryConfigs()
+	if err != nil {
+		log.Error("Error reading registry auth files: %s", err)
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/executor/executor_docker_api.go]: I changed the registry login to only occur if we pull the image which reduces the calls to registry login and as a side effect allows the tests to operate in situations where all the images are local and registry creds are not needed.

File: integration-tests/pkg/log/logging.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+func (l *Logger) Info(format string, v ...interface{}) {
+	if l.level <= InfoLevel {
+		l.logger.Printf("INFO: "+format, v...)
+	}
+}
+
+func (l *Logger) Warn(format string, v ...interface{}) {
+	if l.level <= WarnLevel {
+		l.logger.Printf("WARN: "+format, v...)
+	}
```
[the remark this answers is not in the document]
Comment (robbycochran) [reply integration-tests/pkg/log/logging.go]: Good suggestion. I refactored the logging code

File: integration-tests/suites/k8s_namespace.go (hunk 1 above)
Comment (Molter73) [reply integration-tests/suites/k8s_namespace.go:138]: The sleep is still here though.

**Review by Molter73 — APPROVED** [review]
LGTM!

File: integration-tests/suites/k8s_namespace.go (hunk 1 above)
Comment (robbycochran) [reply integration-tests/suites/k8s_namespace.go]: 🤦 now removed.

**Pull request merged**